File: src/main/java/io/github/juniqlim/realworld/article/repository/rdb/ArticleRDBRepository.java
package io.github.juniqlim.realworld.article.repository.rdb;

import io.github.juniqlim.realworld.Id;
import io.github.juniqlim.realworld.Id.LongId;
import io.github.juniqlim.realworld.article.domain.Article;
import io.github.juniqlim.realworld.article.repository.ArticleRepository;
import org.springframework.data.domain.Sort;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.stream.Collectors;

@Repository
class ArticleRDBRepository implements ArticleRepository {
    private final ArticleJpaRepository articleJpaRepository;
    private final ArticleToArticleEntity articleToArticleEntity = new ArticleToArticleEntity();

    ArticleRDBRepository(ArticleJpaRepository articleJpaRepository) {
        this.articleJpaRepository = articleJpaRepository;
    }

    @Override
    public void save(Article a) {
        articleJpaRepository.save(articleToArticleEntity.articleEntity(a));
    }

    @Override
    public Article findBySlug(String slug) {
        return articleToArticleEntity.article(articleJpaRepository.findBySlug(slug));
    }

    @Override
    public Article findBySlugAndAuthorUserId(String slug, Id authorUserId) {
        return articleToArticleEntity.article(
            articleJpaRepository.findBySlugAndAuthorUserId(slug, authorUserId.value()));
    }

    @Override
    public void update(String slug, Article article) {
        articleJpaRepository.save(articleToArticleEntity.articleEntity(article));
    }

    @Override
    public List<Article> findByRequest(Conditions c) {
        if (!c.ids().isEmpty() && !c.authorUserId().isEmpty()) {
            return articleToArticleEntity.articles(
                articleJpaRepository.findByIdInAndAuthorUserIdOrderByCreatedAt(
                    c.ids().stream()
                        .map(id -> ((LongId) id).value())
                        .collect(Collectors.toList()),
                    c.authorUserId().value(),
                    new OffsetToPage(c.offset(), c.limit(), Sort.by(Direction.DESC, "createdAt"))
                        .pageable()
                )
            );
        }
        if (!c.ids().isEmpty() && c.authorUserId().isEmpty()) {
            return articleToArticleEntity.articles(
                articleJpaRepository.findByIdInOrderByCreatedAt(
                    c.ids().stream()
                        .map(id -> ((LongId) id).value())
                        .collect(Collectors.toList()),
                    new OffsetToPage(c.offset(), c.limit(), Sort.by(Direction.DESC, "createdAt"))
                        .pageable()
                )
            );
        }
        if (c.ids().isEmpty() && !c.authorUserId().isEmpty()) {
            return articleToArticleEntity.articles(
                articleJpaRepository.findAuthorUserIdOrderByCreatedAt(
                    c.authorUserId().value(),
                    new OffsetToPage(c.offset(), c.limit(), Sort.by(Direction.DESC, "createdAt"))
                        .pageable()
                )
            );
        }
        return articleToArticleEntity.articles(
            articleJpaRepository.findByOrderByCreatedAt(
                new OffsetToPage(c.offset(), c.limit(), Sort.by(Direction.DESC, "createdAt"))
                    .pageable()
            )
        );
    }

    @Override
    public void delete(String slug, Id userId) {
        articleJpaRepository.deleteBySlugAndAuthorUserId(slug, userId.value());
    }

    @Override
    public List<Article> findByUserIds(List<Id> followUsers, int offset, int limit) {
        return articleToArticleEntity.articles(
            articleJpaRepository.findByAuthorUserIdIn(
                followUsers.stream()
                    .map(id -> ((LongId)id).value())
                    .collect(Collectors.toList()),
                new OffsetToPage(offset, limit, Sort.by(Direction.DESC, "createdAt"))
                    .pageable()
            )
        );
    }

    @Override
    public Id createId() {
        return new LongId(articleJpaRepository.sequence());
    }
}

File: src/main/kotlin/io/github/juniqlim/realworld/article/web/ArticleRDBRepository.kt
package io.github.juniqlim.realworld.article.web

import io.github.juniqlim.realworld.article.domain.Article
import io.github.juniqlim.realworld.article.repository.ArticleRepository
import io.github.juniqlim.realworld.user.domain.User

class ArticleRDBRepository : ArticleRepository {
    override fun save(article: Article?) {
        TODO("Not yet implemented")
    }

    override fun findBySlug(slug: String?): Article {
        TODO("Not yet implemented")
    }

    override fun findBySlugAndUserId(slug: String?, userId: User.Id?): Article {
        TODO("Not yet implemented")
    }

    override fun update(slug: String?, article: Article?) {
        TODO("Not yet implemented")
    }

    override fun findByTagAuthorIdFavoriteUserIdOrderByRegdate(
        tag: String?,
        authorId: User.Id?,
        favoriteUserId: User.Id?,
        offset: Int,
        limit: Int
    ): MutableList<Article> {
        TODO("Not yet implemented")
    }

    override fun findByTagAuthorFavoriteUserOrderByRegdate(
        tag: String?,
        author: io.github.juniqlim.realworld.user.User?,
        favoriteUser: io.github.juniqlim.realworld.user.User?,
        offset: Int,
        limit: Int
    ): MutableList<Article> {
        TODO("Not yet implemented")
    }

    override fun delete(slug: String?, userId: User.Id?) {
        TODO("Not yet implemented")
    }

    override fun findCommentSequence(): Long {
        TODO("Not yet implemented")
    }

    override fun findByUserIds(followUsers: MutableList<User.Id>?, offset: Int, limit: Int): MutableList<Article> {
        TODO("Not yet implemented")
    }
}